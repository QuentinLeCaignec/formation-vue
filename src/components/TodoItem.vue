<script setup lang="ts">
export interface ITodo {
  text: string;
  done: boolean;
  id: number;
}

interface ITodoItemProps {
  todo: ITodo;
}
const props = defineProps<ITodoItemProps>();

interface ITodoItemEmits {
  (e: "todo:update", todo: ITodo): void;
}
const emits = defineEmits<ITodoItemEmits>();

function handleChecked() {
  emits("todo:update", {...props.todo, done: !props.todo.done})
}
</script>

<template>
  <li :class="['TodoItem', { TodoItem_isDone: todo.done }]">
    <input type="checkbox" :checked="todo.done" @change="handleChecked"/>
    <span>{{ todo.id }} : {{ todo.text }}</span>
  </li>
</template>

<style lang="scss">
.TodoItem {
  padding: 5px;
  background: blanchedalmond;
  border-radius: 5px;
  border: solid 1px gray;

  &_isDone {
    text-decoration-line: line-through;
  }
}
</style>
