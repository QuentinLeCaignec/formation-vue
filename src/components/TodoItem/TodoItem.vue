<script setup lang="ts">
import {nextTick, ref, watch} from 'vue'
import type {ITodo} from '@/types/todo'

type ITodoItemProps = {
  todo: ITodo
}
const props = defineProps<ITodoItemProps>()
type ITodoItemEmits = {
  (e: 'todo:update', todo: ITodo): void
}
const emits = defineEmits<ITodoItemEmits>()

/* Methods */
function handleDone() {
  emits("todo:update", {
    ...props.todo,
    done: !props.todo.done
  })
}

function handleUpdate(e: SubmitEvent) {
  const formData = e.target as HTMLFormControlsCollection;
  emits('todo:update', {
    ...props.todo,
    text: formData["text"].value,
  })
  isEditing.value = false;
}

/* Data */
const isEditing = ref<boolean>(false);
const inputRef = ref<HTMLInputElement>();

watch(isEditing, async () => {
  inputRef.value?.focus();
}, {flush: "post"})

</script>

<template>
  <li class="TodoItem__Container">
    <div :class="['TodoItem', { TodoItem_isDone: todo.done }]">
      <input type="checkbox" :checked="todo.done" @change="handleDone"
             :disabled="isEditing"/>
      <span v-if="!isEditing" data-testid="test-text">{{ todo.text }}</span>
      <form v-else @submit.prevent="handleUpdate">
        <input name="text" :value="todo.text" ref="inputRef" data-testid="test-input"/>
      </form>
    </div>
    <input type="checkbox" v-model="isEditing" data-testid="test-checkbox">
    <slot/>
  </li>
</template>

<style lang="scss">
.TodoItem {
  flex: 1;
  display: flex;
  flex-direction: row;
  gap: 5px;
  padding: 5px;
  background: blanchedalmond;
  border-radius: 5px;
  border: solid 1px gray;
  min-height: 25px;
  align-items: center;

  &_isDone {
    text-decoration-line: line-through;
    background: lightgray;
  }

  &__Container {
    display: flex;
    flex-direction: row;
    gap: 5px;
  }
}
</style>
