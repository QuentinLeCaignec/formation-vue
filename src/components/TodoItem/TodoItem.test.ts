import {describe, it, expect} from "vitest";
import {mount} from "@vue/test-utils";
import TodoItem from "@/components/TodoItem/TodoItem.vue";

const mockTodoItem = {id: 1, text: "mon texte", done: false};

describe('TodoItem', () => {
    const wrapper = mount(TodoItem, {
        props: {todo: mockTodoItem}
    });

    it("renders and displays prop text", () => {
        expect(wrapper.element).toMatchSnapshot();
        expect(wrapper.find("[data-testid='test-text']").text()).toBe("mon texte")
    })
    it("goes into editing mode when right checkbox is clicked", async () => {
        const checkbox = wrapper.get("[data-testid='test-checkbox']");
        await checkbox.setValue(true);
        const input = wrapper.find("[data-testid='test-input']");
        expect(input.exists()).toBe(true);
    })
});
