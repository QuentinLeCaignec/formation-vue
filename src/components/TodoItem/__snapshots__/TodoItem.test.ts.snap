// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TodoItem > renders 1`] = `
<li
  class="TodoItem__Container"
>
  <div
    class="TodoItem"
  >
    <input
      type="checkbox"
    />
    <span
      data-testid="test-text"
    >
      mon texte
    </span>
  </div>
  <input
    data-testid="test-checkbox"
    type="checkbox"
  />
  
  
</li>
`;

exports[`TodoItem > renders and displays prop text 1`] = `
<li
  class="TodoItem__Container"
>
  <div
    class="TodoItem"
  >
    <input
      type="checkbox"
    />
    <span
      data-testid="test-text"
    >
      mon texte
    </span>
  </div>
  <input
    data-testid="test-checkbox"
    type="checkbox"
  />
  
  
</li>
`;
